File: src/rule_interpreter/path_resolver.py
"""
Path resolver for flexible entity attribute access.

Supports:
- Dictionaries (nested dict access)
- Objects (using getattr)
- Dataclasses / Pydantic models
- List indexing (including negative indices)
"""

from dataclasses import fields, is_dataclass
from typing import Any

from rule_interpreter.exceptions import PathResolutionError


def _is_dict_like(obj: Any) -> bool:
    """Check if object supports dict-like access."""
    return isinstance(obj, dict) or hasattr(obj, "__getitem__") and hasattr(obj, "keys")


def _get_attribute(obj: Any, key: str) -> Any:
    """Get an attribute from an object, supporting dicts and objects."""
    # Try dict-like access first for dicts
    if isinstance(obj, dict):
        if key not in obj:
            raise KeyError(key)
        return obj[key]

    # Try attribute access for objects
    if hasattr(obj, key):
        return getattr(obj, key)

    # Try dict-like access for other mappings
    if hasattr(obj, "__getitem__") and hasattr(obj, "keys"):
        try:
            return obj[key]
        except (KeyError, TypeError):
            pass

    raise AttributeError(f"'{type(obj).__name__}' has no attribute '{key}'")


def _get_index(obj: Any, index: int) -> Any:
    """Get an item by index from a sequence."""
    if not hasattr(obj, "__getitem__"):
        raise TypeError(f"'{type(obj).__name__}' does not support indexing")
    return obj[index]


def _set_attribute(obj: Any, key: str, value: Any) -> None:
    """Set an attribute on an object, supporting dicts and objects."""
    if isinstance(obj, dict):
        obj[key] = value
    elif hasattr(obj, "__setitem__") and hasattr(obj, "keys"):
        obj[key] = value
    else:
        setattr(obj, key, value)


def _set_index(obj: Any, index: int, value: Any) -> None:
    """Set an item by index in a sequence."""
    if not hasattr(obj, "__setitem__"):
        raise TypeError(f"'{type(obj).__name__}' does not support index assignment")
    obj[index] = value


class PathResolver:
    """
    Resolves attribute paths on entities.

    Paths can include:
    - Dot notation: entity.user.name
    - Bracket notation for indices: entity.items[0]
    - Negative indices: entity.items[-1]
    """

    def __init__(self, entity: Any, entity_name: str = "entity"):
        """
        Initialize the path resolver.

        Args:
            entity: The root entity to resolve paths on
            entity_name: The name used in path expressions (default: "entity")
        """
        self.entity = entity
        self.entity_name = entity_name
        self._context: dict[str, Any] = {entity_name: entity}

    def add_to_context(self, name: str, value: Any) -> None:
        """Add a variable to the resolution context."""
        self._context[name] = value

    def resolve(self, path_parts: list[str | int]) -> Any:
        """
        Resolve a path to get its value.

        Args:
            path_parts: List of path components (strings for attributes, ints for indices)

        Returns:
            The resolved value

        Raises:
            PathResolutionError: If the path cannot be resolved
        """
        if not path_parts:
            raise PathResolutionError("", type(self.entity).__name__, "Empty path")

        # Get the root from context
        root_name = path_parts[0]
        if not isinstance(root_name, str):
            raise PathResolutionError(
                str(path_parts), type(self.entity).__name__, f"Path must start with an identifier, got {type(root_name)}"
            )

        if root_name not in self._context:
            raise PathResolutionError(str(path_parts), type(self.entity).__name__, f"Unknown identifier '{root_name}'")

        current = self._context[root_name]
        path_str = root_name

        for part in path_parts[1:]:
            try:
                if isinstance(part, int):
                    path_str += f"[{part}]"
                    current = _get_index(current, part)
                else:
                    path_str += f".{part}"
                    current = _get_attribute(current, part)
            except (KeyError, AttributeError, IndexError, TypeError) as e:
                raise PathResolutionError(path_str, type(self.entity).__name__, str(e)) from e

        return current

    def resolve_for_assignment(self, path_parts: list[str | int]) -> tuple[Any, str | int]:
        """
        Resolve a path for assignment, returning the parent object and the final key/index.

        Args:
            path_parts: List of path components

        Returns:
            Tuple of (parent_object, final_key_or_index)

        Raises:
            PathResolutionError: If the path cannot be resolved
        """
        if len(path_parts) < 2:
            raise PathResolutionError(str(path_parts), type(self.entity).__name__, "Cannot assign to root object")

        # Resolve all but the last part
        parent = self.resolve(path_parts[:-1])
        return parent, path_parts[-1]

    def assign(self, path_parts: list[str | int], value: Any) -> None:
        """
        Assign a value to a path.

        Args:
            path_parts: List of path components
            value: The value to assign

        Raises:
            PathResolutionError: If the path cannot be resolved for assignment
        """
        parent, key = self.resolve_for_assignment(path_parts)

        try:
            if isinstance(key, int):
                _set_index(parent, key, value)
            else:
                _set_attribute(parent, key, value)
        except (TypeError, AttributeError) as e:
            path_str = self._path_to_string(path_parts)
            raise PathResolutionError(path_str, type(self.entity).__name__, str(e)) from e

    def _path_to_string(self, path_parts: list[str | int]) -> str:
        """Convert path parts to a string representation."""
        if not path_parts:
            return ""
        result = str(path_parts[0])
        for part in path_parts[1:]:
            if isinstance(part, int):
                result += f"[{part}]"
            else:
                result += f".{part}"
        return result

File: typescript/src/pathResolver.ts
import { PathResolutionError } from "./errors.js";

type PathPart = string | number;

function entityTypeName(value: unknown): string {
  if (value === null) {
    return "null";
  }
  if (value === undefined) {
    return "undefined";
  }
  if (Array.isArray(value)) {
    return "Array";
  }
  if (typeof value === "object") {
    return (value as { constructor?: { name?: string } }).constructor?.name ?? "Object";
  }
  return typeof value;
}

function hasOwnOrInheritedProperty(value: object, key: string): boolean {
  return key in value;
}

function getAttribute(obj: unknown, key: string): unknown {
  if (obj instanceof Map) {
    if (!obj.has(key)) {
      throw new Error(`'Map' has no attribute '${key}'`);
    }
    return obj.get(key);
  }

  if (obj === null || obj === undefined || (typeof obj !== "object" && typeof obj !== "function")) {
    throw new Error(`'${entityTypeName(obj)}' has no attribute '${key}'`);
  }

  if (!hasOwnOrInheritedProperty(obj, key)) {
    throw new Error(`'${entityTypeName(obj)}' has no attribute '${key}'`);
  }

  return (obj as Record<string, unknown>)[key];
}

function normalizeSequenceIndex(length: number, index: number): number {
  return index < 0 ? length + index : index;
}

function getIndex(obj: unknown, index: number): unknown {
  if (obj === null || obj === undefined) {
    throw new Error(`'${entityTypeName(obj)}' does not support indexing`);
  }

  if (typeof obj === "string" || Array.isArray(obj)) {
    const normalizedIndex = normalizeSequenceIndex(obj.length, index);
    if (normalizedIndex < 0 || normalizedIndex >= obj.length) {
      throw new Error(`Index ${index} out of range`);
    }
    return obj[normalizedIndex as keyof typeof obj];
  }

  if (obj instanceof Map) {
    if (!obj.has(index)) {
      throw new Error(`Index ${index} out of range`);
    }
    return obj.get(index);
  }

  if (typeof obj === "object" && index in (obj as { [n: number]: unknown })) {
    return (obj as { [n: number]: unknown })[index];
  }

  throw new Error(`'${entityTypeName(obj)}' does not support indexing`);
}

function setAttribute(obj: unknown, key: string, value: unknown): void {
  if (obj instanceof Map) {
    obj.set(key, value);
    return;
  }

  if (obj === null || obj === undefined || (typeof obj !== "object" && typeof obj !== "function")) {
    throw new Error(`'${entityTypeName(obj)}' has no attribute '${key}'`);
  }

  (obj as Record<string, unknown>)[key] = value;
}

function setIndex(obj: unknown, index: number, value: unknown): void {
  if (obj === null || obj === undefined || typeof (obj as { [n: number]: unknown }) !== "object") {
    throw new Error(`'${entityTypeName(obj)}' does not support index assignment`);
  }

  if (Array.isArray(obj)) {
    const normalizedIndex = normalizeSequenceIndex(obj.length, index);
    if (normalizedIndex < 0 || normalizedIndex >= obj.length) {
      throw new Error(`Index ${index} out of range`);
    }
    obj[normalizedIndex] = value;
    return;
  }

  if (obj instanceof Map) {
    obj.set(index, value);
    return;
  }

  if (!(index in (obj as { [n: number]: unknown }))) {
    throw new Error(`'${entityTypeName(obj)}' does not support index assignment`);
  }

  (obj as { [n: number]: unknown })[index] = value;
}

export class PathResolver {
  public readonly entity: unknown;
  public readonly entityName: string;
  private readonly context: Record<string, unknown>;

  constructor(entity: unknown, entityName = "entity") {
    this.entity = entity;
    this.entityName = entityName;
    this.context = { [entityName]: entity };
  }

  public addToContext(name: string, value: unknown): void {
    this.context[name] = value;
  }

  public normalizePath(pathParts: PathPart[]): PathPart[] {
    if (pathParts.length === 0) {
      return pathParts;
    }

    const [rootName] = pathParts;
    if (typeof rootName === "string" && !(rootName in this.context)) {
      return [this.entityName, ...pathParts];
    }
    return [...pathParts];
  }

  public resolve(pathParts: PathPart[], nullSafeIndices: Set<number> = new Set()): unknown {
    if (pathParts.length === 0) {
      throw new PathResolutionError("", entityTypeName(this.entity), "Empty path");
    }

    const originalLength = pathParts.length;
    const normalized = this.normalizePath(pathParts);
    const shiftedNullSafe =
      normalized.length > originalLength
        ? new Set([...nullSafeIndices].map((index) => index + 1))
        : nullSafeIndices;

    const rootName = normalized[0];
    if (typeof rootName !== "string") {
      throw new PathResolutionError(String(pathParts), entityTypeName(this.entity), "Path must start with an identifier");
    }

    if (!(rootName in this.context)) {
      throw new PathResolutionError(String(pathParts), entityTypeName(this.entity), `Unknown identifier '${rootName}'`);
    }

    let current = this.context[rootName];
    let pathString = rootName;

    for (let offset = 0; offset < normalized.slice(1).length; offset += 1) {
      const part = normalized[offset + 1];
      const segmentIndex = offset + 1;
      const isNullSafe = shiftedNullSafe.has(segmentIndex);

      if (current === null || current === undefined) {
        if (isNullSafe) {
          return null;
        }

        throw new PathResolutionError(pathString, entityTypeName(this.entity), "Cannot access property of None");
      }

      try {
        if (typeof part === "number") {
          pathString += `[${part}]`;
          current = getIndex(current, part);
        } else {
          pathString += `${isNullSafe ? "?." : "."}${part}`;
          current = getAttribute(current, part);
        }
      } catch (error) {
        if (isNullSafe) {
          return null;
        }

        throw new PathResolutionError(
          pathString,
          entityTypeName(this.entity),
          error instanceof Error ? error.message : String(error),
        );
      }
    }

    return current ?? null;
  }

  public resolveForAssignment(pathParts: PathPart[]): [unknown, PathPart] {
    const normalized = this.normalizePath(pathParts);
    if (normalized.length < 2) {
      throw new PathResolutionError(String(pathParts), entityTypeName(this.entity), "Cannot assign to root object");
    }

    return [this.resolve(normalized.slice(0, -1)), normalized[normalized.length - 1]];
  }

  public assign(pathParts: PathPart[], value: unknown): void {
    const [parent, key] = this.resolveForAssignment(pathParts);

    try {
      if (typeof key === "number") {
        setIndex(parent, key, value);
      } else {
        setAttribute(parent, key, value);
      }
    } catch (error) {
      throw new PathResolutionError(
        this.pathToString(pathParts),
        entityTypeName(this.entity),
        error instanceof Error ? error.message : String(error),
      );
    }
  }

  private pathToString(pathParts: PathPart[]): string {
    if (pathParts.length === 0) {
      return "";
    }

    let result = String(pathParts[0]);
    for (const part of pathParts.slice(1)) {
      result += typeof part === "number" ? `[${part}]` : `.${part}`;
    }
    return result;
  }
}
